File: src/main/java/com/leetcode/P2130.kt
package com.leetcode

// https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/
class P2130 {
    fun pairSum(head: ListNode?): Int {
        var prev: ListNode? = null
        var slow = head
        var fast = head
        while (fast != null) {
            fast = fast.next?.next
            val next = slow?.next
            if (slow != null) slow.next = prev
            prev = slow
            slow = next
        }

        var ans = 0
        while (prev != null && slow != null) {
            ans = maxOf(ans, prev.`val` + slow.`val`)
            prev = prev.next
            slow = slow.next
        }
        return ans
    }

    class ListNode(var `val`: Int) {
        var next: ListNode? = null
    }
}

File: src/python3/com/leetcode/p2130.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        prev = None
        slow = head
        fast = head
        while fast is not None:
            fast = fast.next.next
            _next = slow.next
            slow.next = prev
            prev = slow
            slow = _next

        ans = 0
        while prev is not None:
            ans = max(ans, prev.val + slow.val)
            prev = prev.next
            slow = slow.next

        return ans
